### PR TITLE
[MXNET-547] Tutorial explaining how to use the profiler	

### PR DESCRIPTION
**Pull request by indhub**
## Description ##
Tutorial explaining how to use the profiler.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html

### CODE REVIEW

**Comment by indhub** [comment]
@ThomasDelteil @thomelane @Ishitori @safrooze 
Please take a look when you get time.

File: docs/tutorials/python/profiler.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
+# Profiling MXNet Models
+
+It is often helpful to understand what operations take how much time while running a model. This helps optimize the model to run faster. In this tutorial, we will learn how to profile MXNet models to measure their running time and memory consumption using the MXNet profiler.
+
+## The incorrect way to profile
```
Comment (piiswrong) [inline docs/tutorials/python/profiler.md]: This is not incorrect. You can still use wait_to_read to time the dot operation

Comment (indhub) [reply docs/tutorials/python/profiler.md]: Agree. But I don't want to suggest wait_to_read as the recommended way to measure time taken by operations. While it might work for toy problems like this, 
- it is harder to use to measure execution time of multiple operations (requires wait_to_read both before and after the measured operation in multiple places). 
- It is hard to use to measure running time of a block inside a sequence (which is common)
- it won't work for hybrid networks.

The goal of this tutorial is to point people to a recommended way of profiling that works for almost all cases.

However I can add a note along the lines of "While it is possible to use `wait_to_read()` before and after an operation to get running time of an operation, it is not a scalable method to measure running time of multiple operations, especially in a Sequential or Hybrid network"

File: docs/tutorials/python/profiler.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+To use the profiler, you need to build MXNet with `USE_PROFILER` enabled. For example, this command will build the CPU version of MXNet on Linux,
+
+```
+make -j $(nproc) USE_OPENCV=1 USE_BLAS=openblas USE_PROFILER=1
+```
+
+Check [this](http://mxnet.incubator.apache.org/install/index.html?device=Linux&language=Python&processor=CPU) page for more information on building from source for various environments. 
+
+After building with `USE_PROFILER=True` and installing, you can import the profiler and configure it from Python code.
```
Comment (Ishitori) [inline docs/tutorials/python/profiler.md]: USE_PROFILER=True or USE_PROFILER=1 or it doesn't matter?

File: docs/tutorials/python/profiler.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+While it is possible to use `NDArray.waitall()` before and after operations to get running time of operations, it is not a scalable method to measure running time of multiple sets of operations, especially in a Sequential or Hybrid network.
+
+## The correct way to profile
+
+The correct way to measure running time of MXNet models is to use MXNet profiler. In the rest of this tutorial, we will learn how to use the MXNet profiler to measure the running time and memory consumption of MXNet models.
+
+To use the profiler, you need to build MXNet with `USE_PROFILER` enabled. For example, this command will build the CPU version of MXNet on Linux,
+
+```
+make -j $(nproc) USE_OPENCV=1 USE_BLAS=openblas USE_PROFILER=1
```
Comment (Ishitori) [inline docs/tutorials/python/profiler.md]: It would be useful to provide GPU version as well, as I assume many people would want to use it.

File: docs/tutorials/python/profiler.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+```python
+from mxnet.gluon.data.vision import transforms
+train_data = gluon.data.DataLoader(gluon.data.vision.MNIST(train=True).transform_first(transforms.ToTensor()),
+                                   batch_size=64, shuffle=True)
+```
+
+Let's define a method that will run one training iteration given data and label.
+
+```python
+# Use GPU is available
```
Comment (Ishitori) [inline docs/tutorials/python/profiler.md]: Typo? "Use GPU **if** available"

Comment (Ishitori) [reply docs/tutorials/python/profiler.md]: Also, I am not sure that it is safe to include this line. Before you compile CPU version of mxnet and here you actually may end up using gpu() and it will fail.

File: docs/tutorials/python/profiler.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+```
+
+![Profile Statistics](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/tutorials/python/profiler/profile_stats.png)<!--notebook-skip-line-->
+
+#### 2. View in browser
+
+You can also dump the information collected by the profiler into a `json` file using the `profiler.dump()` function and view it in a browser.
+
+```python
+profiler.dump()
```
Comment (Ishitori) [inline docs/tutorials/python/profiler.md]: So, the difference between getting a plain text version vs. json is in calling "dump**s**()" vs "dump()"? Is it possible to change this signature?

Comment (indhub) [reply docs/tutorials/python/profiler.md]: I guess the `s` in `dumps` indicates the method returns a string. Like `pickle.dumps()` or `json.dumps()`

File: docs/tutorials/python/profiler.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+from mxnet.gluon.data.vision import transforms
+train_data = gluon.data.DataLoader(gluon.data.vision.MNIST(train=True).transform_first(transforms.ToTensor()),
+                                   batch_size=64, shuffle=True)
+```
+
+Let's define a method that will run one training iteration given data and label.
+
+```python
+# Use GPU if available
+ctx = mx.gpu() if mx.test_utils.list_gpus() else mx.cpu()
```
Comment (sandeep-krishnamurthy) [inline docs/tutorials/python/profiler.md]: This API has a bug. It doesn't detect GPUs on Windows (due to usage of nvidia-smi command that may not exist in Windows). I have Github issue for this. Pasting here as FYI.
